File: lib/lexer/lexer.c
#include "lexer.h"
#include "helpers.h"
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>
#include <ctype.h>

#define CUR      input[pos]
#define NEXT     input[pos + 1]
#define NEXTNEXT input[pos + 2]
#define PREV     input[pos - 1]

Token make_token(TokenType type, const char* value) {
	return (Token) {
		.type = type,
		.value = value ? strdup(value) : NULL
	};
}

Token* lex(const char* input) {
	Token* tokens = malloc(sizeof(Token) * 128);
	size_t capacity = 128;
	size_t count = 0;  // Token count
	size_t pos = 0;    // Character position in input file

	while (CUR != '\0') {
		// Skip whitespace
		if (isspace(CUR)) {
			pos++;
			continue;
		}
		// Skip one-line comments
		if (CUR == '-' && NEXT == '-') {
			pos += 2;
			while (CUR != '\0' && CUR != '\n') pos++;
			continue;
		}
		// Skip block comments
		if (CUR == '-' && NEXT == '*') {
			pos += 2;
			size_t depth = 1;
			while (CUR != '\0') {
				if (CUR == '-' && NEXT == '*') {
					pos += 2;
					depth++;
				} else if (CUR == '*' && NEXT == '-') {
					pos += 2;
					depth--;
					if (depth == 0) break;
				} else {
					pos++;
				}
			}
			continue;
		}

		// Identifiers & builtin words
		if (isalpha(CUR) || CUR == '_') {
			size_t start = pos;
			while (isalnum(CUR) || CUR == '_') pos++;

			size_t len = pos - start;
			char* word = malloc(len + 1);
			strncpy(word, input + start, len);
			word[len] = '\0';

			TokenType type = get_word_type(word);
			tokens[count++] = make_token(type, type == TOK_IDENT ? word : NULL);
			free(word);
		}

		// Numeric literals
		// TODO: scientific notation
		else if (isdigit(CUR)) {
			size_t start = pos;
			bool has_dot = false;
			TokenType type = TOK_LIT_INT;

			// Hexadecimal
			if (CUR == '0' && (NEXT == 'x' || NEXT == 'X')) {
				pos += 2;
				while (isxdigit(CUR)) pos++;
			} // Octal
			else if (CUR == '0' && (NEXT == 'o' || NEXT == 'O')) {
				pos += 2;
				while (CUR >= '0' && CUR <= '7') pos++;
			} // Binary
			else if (CUR == '0' && (NEXT == 'b' || NEXT == 'B')) {
				pos += 2;
				while (CUR == '0' || CUR == '1') pos++;
			} // Decimal
			else {
				while (isdigit(CUR) ||
						(CUR == '.' && !has_dot && isdigit(NEXT))) {
					if (CUR == '.') has_dot = true;
					pos++;
				}
				if (has_dot) type = TOK_LIT_FLOAT;
			}

			size_t len = pos - start;
			char* num = malloc(len + 1);
			strncpy(num, input + start, len);
			num[len] = '\0';

			tokens[count++] = make_token(type, num);
			free(num);
		}

		// String literals
		else if (CUR == '"' || CUR == '`') {
			pos++;  // Skip opening quote
			size_t start = pos;
			TokenType type = PREV == '"' ? TOK_LIT_STR : TOK_LIT_STR_RAW;

			while (CUR != '\0' && CUR != (type == TOK_LIT_STR ? '"' : '`')) {
				if (CUR == '\\') pos++;
				if (CUR == '\0') break;
				pos++;
			}
			if (CUR == '\0') {
				tokens[count++] = make_token(TOK_ERR, "Unterminated string literal");
				continue;
			}

			size_t len = pos - start;
			char* str = malloc(len + 1);
			strncpy(str, input + start, len);
			str[len] = '\0';

			tokens[count++] = make_token(type, str);
			free(str);
			pos++;  // Skip closing quote
		}

		// Character literals
		else if (CUR == '\'') {
			pos++;  // Skip opening quote
			char ch[3];

			if (CUR != '\0' && NEXT != '\0') {
				if (CUR == '\\' && NEXTNEXT == '\'') {
					ch[0] = '\\';
					ch[1] = NEXT;
					ch[2] = '\0';
					pos += 2;
				} else if (CUR != '\'' && NEXT == '\'') {
					ch[0] = CUR;
					ch[1] = '\0';
					ch[2] = '\0';
					pos++;
				} else {
					while (CUR != '\0' && CUR != '\'') pos++;
					if (CUR == '\'') pos++;
					tokens[count++] = make_token(TOK_ERR, "Invalid character literal");
					continue;
				}
			} else {
				tokens[count++] = make_token(TOK_ERR, "Unterminated character literal");
				if (CUR != '\0') pos++;
				continue;
			}

			tokens[count++] = make_token(TOK_LIT_CHAR, ch);
			pos++;  // Skip closing quote
		}

		// Punctuation & operators
		else {
			TokenType type = get_symbol_type(CUR, NEXT);
			if (is_long_symbol(type)) pos++;
			pos++;
			if (type == TOK_ERR) {
				char err_str[32];
				sprintf(err_str, "Unknown character %c", PREV);
				tokens[count++] = make_token(TOK_ERR, err_str);
			} else {
				tokens[count++] = make_token(type, NULL);
			}
		}

		// Resize the token array if needed
		if (count >= capacity) {
			capacity *= 2;
			tokens = realloc(tokens, sizeof(Token) * capacity);
		}
	}

	// Safely add EOF token
	if (count >= capacity) {
		capacity++;
		tokens = realloc(tokens, sizeof(Token) * capacity);
	}
	tokens[count++] = make_token(TOK_EOF, NULL);

	return tokens;
}

void free_tokens(Token* tokens) {
	if (!tokens) return;
	for (int i = 0; tokens[i].type != TOK_EOF; i++) {
		if (tokens[i].value) free(tokens[i].value);
	}
	free(tokens);
}

File: src/lexer/lexer.rs
use crate::lexer::token::Token;
use crate::parser::parser::ParserError;

pub struct Lexer {
	chars: Vec<char>,
	pos: usize,
	pub current: Option<Token>,
}

impl Lexer {
	pub fn new(input: &str) -> Self {
		Lexer {
			chars: input.chars().collect(),
			pos: 0,
			current: None,
		}
	}

	pub fn consume(&mut self) -> Option<Token> {
		self.current = self.lex_token();
		self.current.clone()
	}

	// Step backwards until we find a semicolon, `end` or `else`
	pub fn goto_last_terminator(&mut self) -> Result<(), ParserError> {
		while self.pos > 0 {
			match self.peek(0).unwrap() {
				// Found semicolon terminator
				';' => return Ok(()),

				// Check if this is an `end` or `else`
				'e' => {
					let prev = self.peek(-1).unwrap();
					// Continue if the e is in the middle of the word
					if prev.is_alphanumeric() || prev == '_' { continue; }
					// Record start position so we can go back to it
					let start = self.pos;
					match self.read_identifier().as_str() {
						"end" | "else" => return Ok(()),
						_ => { self.pos = start; },
					}
				},

				_ => {},
			}
			// Unconditionally step backwards
			self.pos -= 1;
		}
		Err(ParserError::LastTerminatorNotFound)
		// I think this error is unreachable... I think
	}

	fn peek(&self, offset: isize) -> Option<char> {
		let peek_pos = (self.pos as isize + offset) as usize;
		if peek_pos < self.chars.len() { Some(self.chars[peek_pos]) }
		else { None }
	}

	fn advance(&mut self) -> Option<char> {
		self.pos += 1;
		self.peek(0)
	}

	fn read_identifier(&mut self) -> String {
		let start = self.pos;
		while let Some(c) = self.advance() {
			if !c.is_alphanumeric() && c != '_' { break; }
		}
		self.chars[start..self.pos].iter().collect()
	}

	fn read_number(&mut self) -> String {
		let start = self.pos;
		let cur = self.peek(0).unwrap();
		let next = self.peek(1);

		// Check for 0x, 0o, or 0b prefix
		let base = match (cur, next) {
			('0', Some('x')) | ('0', Some('X')) => { self.advance(); 16 },
			('0', Some('o')) | ('0', Some('O')) => { self.advance(); 8 },
			('0', Some('b')) | ('0', Some('B')) => { self.advance(); 2 },
			_ => 10,
		};
		// Read integral part
		while let Some(c) = self.advance() {
			if !c.is_digit(base) { break; }
		}

		if base == 10 {
			// Fractional part
			if let Some('.') = self.peek(0) {
				while let Some(c) = self.advance() {
					if !c.is_digit(10) { break; }
				}
			}
			// Scientific notation
			if let Some('e' | 'E') = self.peek(0) {
				if let Some('+' | '-') = self.advance() {
					self.advance();
				}
				while let Some(c) = self.peek(0) {
					if !c.is_digit(10) { break; }
					self.advance();
				}
			}
		}

		self.chars[start..self.pos].iter().collect()
	}

	fn read_textual_literal(&mut self) -> String {
		let start = self.pos + 1;  // Skip opening quote
		let is_string = self.peek(0).unwrap() == '"';

		while let Some(c) = self.advance() {
			if c == '\\' {
				self.advance();  // Skip escaped character
			} else if (c == '"' && is_string) || (c == '\'' && !is_string) {
				self.advance();
				break;
			}
		}

		let end = self.pos - 1;  // Skip closing quote
		if start + 1 < end { self.chars[start..end].iter().collect() }
		else { String::new() }
	}

	fn skip_whitespace(&mut self) {
		while let Some(c) = self.peek(0) {
			if !c.is_whitespace() { break; }
			self.advance();
		}
	}

	fn lex_token(&mut self) -> Option<Token> {
		self.skip_whitespace();
		let current = self.peek(0)?;

		// Identifiers and keywords
		if current.is_alphabetic() || current == '_' {
			let word = self.read_identifier();
			match word.as_str() {
				// Check for keyword {{{
				"let" => Some(Token::WordLet),
				"return" => Some(Token::WordReturn),
				"if" => Some(Token::WordIf),
				"then" => Some(Token::WordThen),
				"else" => Some(Token::WordElse),
				"elif" => Some(Token::WordElif),
				"case" => Some(Token::WordCase),
				"of" => Some(Token::WordOf),
				"end" => Some(Token::WordEnd),
				"while" => Some(Token::WordWhile),
				"for" => Some(Token::WordFor),
				"in" => Some(Token::WordIn),
				"where" => Some(Token::WordWhere),
				"do" => Some(Token::WordDo),
				"done" => Some(Token::WordDone),
				"break" => Some(Token::WordBreak),
				"continue" => Some(Token::WordContinue),
				"match" => Some(Token::WordMatch),
				"test" => Some(Token::WordTest),
				"const" => Some(Token::WordConst),
				"type" => Some(Token::WordType),
				"def" => Some(Token::WordDef),
				"proc" => Some(Token::WordProc),
				"fn" => Some(Token::WordFn),
				"record" => Some(Token::WordRecord),
				"union" => Some(Token::WordUnion),
				"trait" => Some(Token::WordTrait),
				"impl" => Some(Token::WordImpl),
				"and" => Some(Token::WordAnd),
				"or" => Some(Token::WordOr),
				"xor" => Some(Token::WordXor),
				"not" => Some(Token::WordNot),
				// }}}
				_ => Some(Token::Identifier(word)),
			}
		}

		// Numbers
		else if current.is_digit(10) {
			let num_str = self.read_number();

			// Determine the numeric base (16, 8, 2, or 10)
			let (base, num_body) = if num_str.len() > 1 {
				match &num_str[..2] {
					"0x" | "0X" => (16, &num_str[2..]),
					"0o" | "0O" => (8, &num_str[2..]),
					"0b" | "0B" => (2, &num_str[2..]),
					_ => (10, &num_str[..]),
				}
			} else {
				(10, &num_str[..])
			};

			// Make float or int literal token
			if num_str.contains('.')
			|| num_str.contains('e')
			|| num_str.contains('E') {
				Some(Token::FloatLiteral(num_str))
			} else if let Ok(int_value) = u64::from_str_radix(num_body, base) {
				Some(Token::IntLiteral(int_value))
			} else {
				panic!(
					"Integer literal badly formatted or out of range: {}",
					num_str
				);
			}
		}

		// String and character literals
		else if current == '"' {
			Some(Token::StringLiteral(self.read_textual_literal()))
		} else if current == '\'' {
			Some(Token::CharLiteral(self.read_textual_literal()))
		}

		// Symbols and comments
		else {
			let next = self.peek(1);

			match (current, next) {
				// Skip line comments
				('-', Some('-')) => {
					self.advance();
					while let Some(c) = self.advance() {
						if c == '\n' { break; }
					}
					self.lex_token()
				},

				// Skip block comments
				('-', Some('*')) => {
					self.advance();
					self.advance();
					let mut depth = 1;
					while let (Some(cur), Some(nxt)) = (self.peek(0), self.peek(1)) {
						self.advance();
						match (cur, nxt) {
							('-', '*') => {
								self.advance();
								depth += 1;
							},
							('*', '-') => {
								self.advance();
								depth -= 1;
								if depth == 0 { break; }
							},
							_ => {},
						}
					}
					self.lex_token()
				},

				_ => {
					// Handle two- and one-character symbols
					let (symbol, length) = match (current, next, self.peek(2)) {
						// {{{
						('.', Some('.'), Some('=')) => (Some(Token::InclusiveRange), 3),
						('+', Some('+'), Some('=')) => (Some(Token::ConcatAssign), 3),
						('.', Some('.'), _) => (Some(Token::ExclusiveRange), 2),
						('+', Some('+'), _) => (Some(Token::Concat), 2),
						('-', Some('>'), _) => (Some(Token::Arrow), 2),
						('=', Some('>'), _) => (Some(Token::FatArrow), 2),
						('|', Some('|'), _) => (Some(Token::DoublePipe), 2),
						('=', Some('='), _) => (Some(Token::TestEq), 2),
						('!', Some('='), _) => (Some(Token::TestNe), 2),
						('<', Some('='), _) => (Some(Token::CompLe), 2),
						('>', Some('='), _) => (Some(Token::CompGe), 2),
						('>', Some('>'), _) => (Some(Token::FwdCompose), 2),
						(':', Some(':'), _) => (Some(Token::ModuleAccess), 2),
						('+', Some('='), _) => (Some(Token::AddAssign), 2),
						('-', Some('='), _) => (Some(Token::SubAssign), 2),
						('*', Some('='), _) => (Some(Token::MulAssign), 2),
						('/', Some('='), _) => (Some(Token::DivAssign), 2),
						('%', Some('='), _) => (Some(Token::ModAssign), 2),
						('|', Some('='), _) => (Some(Token::InfixAssign), 2),
						('(', _, _) => (Some(Token::ParenL), 1),
						(')', _, _) => (Some(Token::ParenR), 1),
						('[', _, _) => (Some(Token::BracketL), 1),
						(']', _, _) => (Some(Token::BracketR), 1),
						('{', _, _) => (Some(Token::BraceL), 1),
						('}', _, _) => (Some(Token::BraceR), 1),
						('.', _, _) => (Some(Token::Dot), 1),
						(',', _, _) => (Some(Token::Comma), 1),
						(';', _, _) => (Some(Token::Semicolon), 1),
						(':', _, _) => (Some(Token::Colon), 1),
						('=', _, _) => (Some(Token::Equals), 1),
						('|', _, _) => (Some(Token::Pipe), 1),
						('!', _, _) => (Some(Token::Bang), 1),
						('&', _, _) => (Some(Token::Ampersand), 1),
						('#', _, _) => (Some(Token::Hash), 1),
						('^', _, _) => (Some(Token::Caret), 1),
						('~', _, _) => (Some(Token::Tilde), 1),
						('-', _, _) => (Some(Token::Minus), 1),
						('+', _, _) => (Some(Token::Plus), 1),
						('*', _, _) => (Some(Token::Star), 1),
						('/', _, _) => (Some(Token::Slash), 1),
						('%', _, _) => (Some(Token::Percent), 1),
						('<', _, _) => (Some(Token::Less), 1),
						('>', _, _) => (Some(Token::Greater), 1),
						// }}}
						_ => (None, 0),
					};

					for _ in 0..length { self.advance(); }

					match symbol {
						Some(sym) => Some(sym),
						_ => {
							let c = self.peek(0).unwrap_or('\0');
							self.advance();
							Some(Token::Unknown(c))
						},
					}
				},
			}
		}
	}
}
